### ISSUE TITLE
Duplicate HTTP Response Header for X-Kong-Limit cause error in response-ratelimiting plugin

### ISSUE THREAD

**Issue by bstick12** [Mashape/kong#2089]
### Summary

Duplicate HTTP Response Header for X-Kong-Limit cause error in response-ratelimiting plugin

### Steps To Reproduce
1. Enable the reponse-ratelimiting plugin
2. Have an upstream service return multiple header values for 'X-Kong-Plugin' e.g.
```
< Server: Apache-Coyote/1.1
< X-Application-Context: bootstrap:11500
< X-Kong-Limit: IDRequest=1
< X-Kong-Limit: IDAllocation=9
< Content-Type: application/json;charset=UTF-8
< Transfer-Encoding: chunked
< Date: Tue, 14 Feb 2017 16:14:50 GMT
```
3. You will get a `curl: (52) Empty reply from server`
4. The error.log of kong contains the following
```
/usr/local/share/lua/5.1/pl/stringx.lua:165: in function 'split'
...5.1/kong/plugins/response-ratelimiting/header_filter.lua:17: in function 'parse_header'
...5.1/kong/plugins/response-ratelimiting/header_filter.lua:39: in function 'execute'
...e/lua/5.1/kong/plugins/response-ratelimiting/handler.lua:21: in function 'header_filter'

```
5. According to https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2 having multiple headers with the same name is acceptable

### Additional Details & Logs

- Kong version 0.9.8
- Kong debug-level startup logs 
```
2017/02/14 15:55:03 [verbose] Kong: 0.9.8
2017/02/14 15:55:03 [debug] ngx_lua: 10006
2017/02/14 15:55:03 [debug] nginx: 1011002
2017/02/14 15:55:03 [debug] Lua: LuaJIT 2.1.0-beta2
2017/02/14 15:55:03 [debug] PRNG seed: 182151361912
2017/02/14 15:55:03 [verbose] no config file found at /etc/kong.conf
2017/02/14 15:55:03 [verbose] no config file found at /etc/kong/kong.conf
2017/02/14 15:55:03 [verbose] no config file, skipping loading
2017/02/14 15:55:03 [debug] KONG_CASSANDRA_CONTACT_POINTS ENV found with "kong-database"
2017/02/14 15:55:03 [debug] KONG_PG_HOST ENV found with "kong-database"
2017/02/14 15:55:03 [debug] KONG_DATABASE ENV found with "postgres "
2017/02/14 15:55:03 [debug] admin_listen = "0.0.0.0:8001"
2017/02/14 15:55:03 [debug] anonymous_reports = true
2017/02/14 15:55:03 [debug] cassandra_consistency = "ONE"
2017/02/14 15:55:03 [debug] cassandra_contact_points = {"kong-database"}
2017/02/14 15:55:03 [debug] cassandra_data_centers = {"dc1:2","dc2:3"}
2017/02/14 15:55:03 [debug] cassandra_keyspace = "kong"
2017/02/14 15:55:03 [debug] cassandra_port = 9042
2017/02/14 15:55:03 [debug] cassandra_repl_factor = 1
2017/02/14 15:55:03 [debug] cassandra_repl_strategy = "SimpleStrategy"
2017/02/14 15:55:03 [debug] cassandra_ssl = false
2017/02/14 15:55:03 [debug] cassandra_ssl_verify = false
2017/02/14 15:55:03 [debug] cassandra_timeout = 5000
2017/02/14 15:55:03 [debug] cassandra_username = "kong"
2017/02/14 15:55:03 [debug] cluster_listen = "0.0.0.0:7946"
2017/02/14 15:55:03 [debug] cluster_listen_rpc = "127.0.0.1:7373"
2017/02/14 15:55:03 [debug] cluster_profile = "wan"
2017/02/14 15:55:03 [debug] cluster_ttl_on_failure = 3600
2017/02/14 15:55:03 [debug] custom_plugins = {}
2017/02/14 15:55:03 [debug] database = "postgres"
2017/02/14 15:55:03 [debug] dnsmasq = true
2017/02/14 15:55:03 [debug] dnsmasq_port = 8053
2017/02/14 15:55:03 [debug] log_level = "notice"
2017/02/14 15:55:03 [debug] lua_code_cache = "on"
2017/02/14 15:55:03 [debug] lua_package_cpath = ""
2017/02/14 15:55:03 [debug] lua_package_path = "?/init.lua;./kong/?.lua"
2017/02/14 15:55:03 [debug] lua_ssl_verify_depth = 1
2017/02/14 15:55:03 [debug] mem_cache_size = "128m"
2017/02/14 15:55:03 [debug] nginx_daemon = "on"
2017/02/14 15:55:03 [debug] nginx_optimizations = true
2017/02/14 15:55:03 [debug] nginx_worker_processes = "auto"
2017/02/14 15:55:03 [debug] pg_database = "kong"
2017/02/14 15:55:03 [debug] pg_host = "kong-database"
2017/02/14 15:55:03 [debug] pg_port = 5432
2017/02/14 15:55:03 [debug] pg_ssl = false
2017/02/14 15:55:03 [debug] pg_ssl_verify = false
2017/02/14 15:55:03 [debug] pg_user = "kong"
2017/02/14 15:55:03 [debug] prefix = "/usr/local/kong/"
2017/02/14 15:55:03 [debug] proxy_listen = "0.0.0.0:8000"
2017/02/14 15:55:03 [debug] proxy_listen_ssl = "0.0.0.0:8443"
2017/02/14 15:55:03 [debug] serf_path = "serf"
2017/02/14 15:55:03 [debug] ssl = true
2017/02/14 15:55:03 [verbose] prefix in use: /usr/local/kong
2017/02/14 15:55:03 [verbose] preparing nginx prefix directory at /usr/local/kong
2017/02/14 15:55:03 [verbose] saving serf identifier to /usr/local/kong/serf/serf.id
2017/02/14 15:55:03 [debug] searching for OpenResty 'resty' executable
2017/02/14 15:55:03 [debug] /usr/local/openresty/bin/resty -V: 'nginx version: openresty/1.11.2.1'
2017/02/14 15:55:03 [debug] found OpenResty 'resty' executable at /usr/local/openresty/bin/resty
2017/02/14 15:55:03 [verbose] saving serf shell script handler to /usr/local/kong/serf/serf_event.sh
2017/02/14 15:55:03 [verbose] SSL enabled, no custom certificate set: using default certificate
2017/02/14 15:55:03 [verbose] default SSL certificate found at /usr/local/kong/ssl/kong-default.crt
2017/02/14 15:55:03 [verbose] running datastore migrations
2017/02/14 15:55:03 [verbose] migrations up to date
2017/02/14 15:55:03 [debug] sending signal to pid at: /usr/local/kong/pids/dnsmasq.pid
2017/02/14 15:55:03 [debug] kill -0 `cat /usr/local/kong/pids/dnsmasq.pid` >/dev/null 2>&1
2017/02/14 15:55:03 [verbose] dnsmasq already running at /usr/local/kong/pids/dnsmasq.pid
2017/02/14 15:55:03 [debug] sending signal to pid at: /usr/local/kong/pids/serf.pid
2017/02/14 15:55:03 [debug] kill -0 `cat /usr/local/kong/pids/serf.pid` >/dev/null 2>&1
2017/02/14 15:55:03 [verbose] serf agent already running at /usr/local/kong/pids/serf.pid
2017/02/14 15:55:03 [debug] searching for OpenResty 'nginx' executable
2017/02/14 15:55:03 [debug] /usr/local/openresty/nginx/sbin/nginx -v: 'nginx version: openresty/1.11.2.1'
2017/02/14 15:55:03 [debug] found OpenResty 'nginx' executable at /usr/local/openresty/nginx/sbin/nginx
2017/02/14 15:55:03 [debug] sending signal to pid at: /usr/local/kong/pids/nginx.pid
2017/02/14 15:55:03 [debug] kill -0 `cat /usr/local/kong/pids/nginx.pid` >/dev/null 2>&1
2017/02/14 15:55:03 [verbose] could not start Kong, stopping services
2017/02/14 15:55:03 [debug] sending signal to pid at: /usr/local/kong/pids/nginx.pid
2017/02/14 15:55:03 [debug] kill -0 `cat /usr/local/kong/pids/nginx.pid` >/dev/null 2>&1
2017/02/14 15:55:03 [verbose] sending TERM signal to nginx running at /usr/local/kong/pids/nginx.pid
2017/02/14 15:55:03 [debug] sending signal to pid at: /usr/local/kong/pids/nginx.pid
2017/02/14 15:55:03 [debug] kill -s TERM `cat /usr/local/kong/pids/nginx.pid` >/dev/null 2>&1
2017/02/14 15:55:03 [verbose] leaving serf cluster
```

- Kong error logs 

```
2017/02/14 15:59:04 [error] 79#0: *15 failed to run header_filter_by_lua*: /usr/local/share/lua/5.1/pl/utils.lua:173: argument 1 expected a 'string', got a 'table'
stack traceback:
        [C]: in function 'error'
        /usr/local/share/lua/5.1/pl/utils.lua:388: in function 'assert_arg'
        /usr/local/share/lua/5.1/pl/utils.lua:400: in function 'assert_string'
        /usr/local/share/lua/5.1/pl/utils.lua:173: in function 'usplit'
        /usr/local/share/lua/5.1/pl/stringx.lua:165: in function 'split'
        ...5.1/kong/plugins/response-ratelimiting/header_filter.lua:17: in function 'parse_header'
        ...5.1/kong/plugins/response-ratelimiting/header_filter.lua:39: in function 'execute'
        ...e/lua/5.1/kong/plugins/response-ratelimiting/handler.lua:21: in function 'header_filter'
        /usr/local/share/lua/5.1/kong.lua:174: in function 'header_filter'
        header_filter_by_lua:2: in function <header_filter_by_lua:1> while reading response header from upstream, client: 192.168.99.1, server: kong, request: "GET

```

- Kong configuration (registered APIs/Plugins & configuration file)
```
response-ratelimit plugin
```

- Operating System
```
Kong docker container
```

**Comment by thefosk** [comment]
This is a bug caused by the fact that two headers with the same name return a table with the values. The options we have are:

* We only pick the last header
* We only pick the first header
* We don't pick any header and log an error

**Comment by bstick12** [comment]
@thefosk There is another option which I believe is more correct. According to the [RFC2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2). 

> Multiple message-header fields with the same field-name MAY be present in a message if and only if the entire field-value for that header field is defined as a comma-separated list [i.e., #(values)]. It MUST be possible to combine the multiple header fields into one "field-name: field-value" pair, without changing the semantics of the message, by appending each subsequent field-value to the first, each separated by a comma.

So the values of the headers should be appended to each other using a comma. So

```
X-Kong-Limit: IDRequest=1
X-Kong-Limit: IDAllocation=9
```
Becomes
```
X-Kong-Limit: IDRequest=1, IDAllocation=9
```
Which can be processed using the current mechanism

**Comment by thefosk** [comment]
@bstick12 you are right, that's actually the best solution.

**Comment by jdmonin** [comment]
@thefosk Would you rather have that header consolidation done in core or the rate-limiting plugins?

**Comment by thefosk** [comment]
@jdmonin the fix needs to be implemented [here](https://github.com/Mashape/kong/blob/master/kong/plugins/response-ratelimiting/header_filter.lua#L14).

`header_value` is assumed to be a string value, but sometimes can be a table.

**Comment by jdmonin** [comment]
@thefosk Thank you! I was able to reproduce the issue and sent PR #2108.

**Comment by thibaultcha** [comment]
Just merged #2108 in both release/0.10.0 and master. Expect it to land in 0.9.10 or 0.10.0. Thanks for the report!
